modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityHCPDislocationSlipBeyerleinUpdate.C

```diff
@@ -0,0 +1,602 @@
+        "Please ensure that the size of slip_generation_coefficient_per_mode equals the value "
+        "supplied for slip_system_modes");
+
+  if (_slip_activation_energy.size() != _slip_system_modes)
+    mooseError("Please ensure that the size of normalized_slip_activiation_energy_per_mode equals "
+               "the value supplied for slip_system_modes");
+
+  if (_proportionality_factor.size() != _slip_system_modes)
+    mooseError("Please ensure that the size of slip_energy_proportionality_factor_per_mode equals "
+               "the value supplied for slip_system_modes");
+
+  if (_shear_modulus.size() != _slip_system_modes)
+    mooseError("Please ensure that the size of effective_shear_modulus_per_mode equals the value "
+               "supplied for slip_system_modes");
+
+  if (_substructure_rate_coefficient.size() != _slip_system_modes)
+    mooseError(
+        "Please ensure that the size of substructure_rate_coefficient_per_mode equals the value "
+        "supplied for slip_system_modes");
+
+  if (_hallpetch_like_coefficient.size() != _slip_system_modes)
+    mooseError("Please ensure that the size of Hall_Petch_like_constant_per_mode equals the value "
+               "supplied for slip_system_modes");
+
+  if (_lattice_friction.size() != _slip_system_modes)
+    mooseError("Error in accessing the entries of the lattice_friction_per_mode vector. "
+               "Please ensure that the size of lattice_friction_per_mode equals the value "
+               "supplied for slip_system_modes");
+
+  CrystalPlasticityStressUpdateBase::initQpStatefulProperties();
+
+  // Resize constitutive-model specific material properties
+  _forest_dislocation_density[_qp].resize(_number_slip_systems);
+  _forest_dislocation_increment[_qp].resize(_number_slip_systems);
+  _forest_dislocations_removed_increment[_qp].resize(_number_slip_systems);
+  _initial_lattice_friction.resize(_number_slip_systems);
+
+  // Set constitutive-model specific initial values from parameters
+  const Real forest_density_per_system = _initial_forest_dislocation_density / _number_slip_systems;
+  for (unsigned int i = 0; i < _number_slip_systems; ++i)
+  {
+    _forest_dislocation_density[_qp][i] = forest_density_per_system;
+    _forest_dislocation_increment[_qp][i] = 0.0;
+    _slip_increment[_qp][i] = 0.0;
+  }
+
+  // Set initial resistance from lattice friction, which is type dependent
+  unsigned int slip_mode = 0;
+  unsigned int counter_adjustment = 0;
+  for (unsigned int i = 0; i < _number_slip_systems; ++i)
+  {
+    if ((i - counter_adjustment) < _number_slip_systems_per_mode[slip_mode])
+      _initial_lattice_friction(i) = _lattice_friction[slip_mode];
+    else
+    {
+      counter_adjustment += _number_slip_systems_per_mode[slip_mode];
+      ++slip_mode;
+      mooseAssert(slip_mode < _slip_system_modes,
+                  "Error in accessing the number of slip systems provided for "
+                  "number_slip_systems_per_mode");
```

Comment:
Is this necessary with all the parameter checks above? I am not sure if we can avoid some code duplication here and later?

Reply:
It's a good question, and I'm not sure. I did have to switch from using `mooseError` to `mooseAssert` here because I couldn't figure out how to write a test to hit this error...I'd like to have _something_ here to catch potential issues with the code logic, but there's probably a better way to do so. I'll keep working on this part

Reply:
At the end, I decided you were right that the error checks above are enough and removed all these asserts (that weren't all that helpful in the first place)